src/Core/Settings.h

```diff
@@ -845,6 +845,7 @@ class IColumn;
     M(UInt64, cache_warmer_threads, 4, "Only available in ClickHouse Cloud", 0) \
     M(Int64, ignore_cold_parts_seconds, 0, "Only available in ClickHouse Cloud", 0) \
     M(Int64, prefer_warmed_unmerged_parts_seconds, 0, "Only available in ClickHouse Cloud", 0) \
+    M(Bool, enable_order_by_all, true, "Enable sorting expression ORDER BY ALL.", 0)\
```

Comment:
I'm curious, why do we need a setting for this? Do we really expect for it to ever be disabled?

Reply:
Sorry for the late reply (due to holidays): The settings exists to disambiguate `ORDER BY ALL` when the table has a column named `ALL`. Example from the [docs](https://github.com/ClickHouse/ClickHouse/pull/57875/files#diff-5ddb9754e75d544caa4b0fd9d1889e15ae8cd9721f974883b637be679ec533d6): https://fiddle.clickhouse.com/20f45646-10e8-418a-89de-446a7d6ef47a

Reply:
The naming of this setting is misleading. It should have been called something like `prefer_column_name_for_order_by_all`. And that should affect on what mode we **prefer**.
The following query works with any value of `enable_order_by_all` setting:
```sql
SELECT * FROM TAB ORDER BY ALL, ALL
```
Also, QueryNode is modified in this PR incorrectly, because we need to update `clone`, `toAST`, etc. I think this feature is poorly designed.

Reply:
> The naming of this setting is misleading. It should have been called something like prefer_column_name_for_order_by_all. And that should affect on what mode we prefer.

This is about the unlikely edge case that the result set already has a column/alias/expression named `all`. The current logic throws an exception when the user runs `ORDER BY ALL` (or `ORDER BY all`) in such a situation. I think this is less error-prone than silently preferring one alternative over the other.

About the setting name: We already have a setting `enable_positional_arguments` which affects `ORDER BY`, `GROUP BY` and `LIMIT BY`. I guess for consistency, we should
- rename setting `enable_order_by_all` to `enable_all_argument`, and
- additionally implement support for `ALL` in `GROUP BY` and `LIMIT BY` (this will be nice-to-have, I can open an issue for that)

As mentioned, I'd rather not change the semantics of the setting to avoid unpleasant surprises.

> The following query works with any value of enable_order_by_all setting:
> SELECT * FROM TAB ORDER BY ALL, ALL

Interestingly, this situation was kind of anticipated and there is [a test](https://github.com/ClickHouse/ClickHouse/blob/a8845ba729224ab10e21349a416ec5f3b5d84eab/tests/queries/0_stateless/02943_order_by_all.sql#L84) for it. Basically, when `ALL` is not used stand-alone, it currently loses its special meaning. I agree that we should forbid this situation, i.e. throw an exception if `ALL` is used together with other expressions (with default setting of `enable_order_by_all = 1`) and let the user to disable the setting. This is also the way DuckDB implements it ([docs](https://duckdb.org/docs/sql/query_syntax/orderby.html), "Note that ALL may not be used in combination with other expressions in the ORDER BY clause - it must be by itself.").

Also, QueryNode is modified in this PR incorrectly, because we need to update clone, toAST, etc. I think this feature is poorly designed.

> Also, QueryNode is modified in this PR incorrectly, because we need to update clone, toAST, etc.

I don't know the Analyzer code too much yet. What do you exactly mean is wrong with the [code](https://github.com/ClickHouse/ClickHouse/blob/0791c75315f6ac50ec38f3f884f32ad1157f4f8f/src/Analyzer/Passes/QueryAnalysisPass.cpp#L2331)? Are there similar transformations for inspiration?